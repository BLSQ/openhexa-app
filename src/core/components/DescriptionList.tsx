import { InformationCircleIcon } from "@heroicons/react/20/solid";
import clsx from "clsx";
import React, { ReactNode } from "react";
import Tooltip from "./Tooltip";

const DescriptionList = ({
  children,
  className,
}: React.HTMLAttributes<HTMLDataListElement>) => {
  return <dl className={clsx("space-y-6", className)}>{children}</dl>;
};

DescriptionList.Item = function Item({
  children,
  label,
  help,
  className,
}: {
  label: ReactNode;
  children: ReactNode;
  help?: ReactNode;
  className?: string;
}) {
  return (
    <div className={clsx("grid grid-cols-5 gap-2 sm:gap-4", className)}>
      <dt className="col-span-1 text-sm font-medium text-gray-500">
        <span>{label}</span>
        {help && (
          <Tooltip
            placement="top"
            renderTrigger={(ref) => (
              <span ref={ref}>
                <InformationCircleIcon className="ml-1 h-3 w-3 cursor-pointer" />
              </span>
            )}
            label={help}
          />
        )}
      </dt>
      <dd className="col-span-4 text-sm text-gray-900">{children}</dd>
    </div>
  );
};

export default DescriptionList;
