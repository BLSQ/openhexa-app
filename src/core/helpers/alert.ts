export enum AlertType {
  success,
  error,
  info,
  warning,
}

export function displayAlert(message: string, type?: AlertType) {
  if (!window) {
    console.warn("displayAlert cannot be called on the server.");
    return;
  }
  const event = new CustomEvent("displayAlert", {
    detail: {
      message,
      type: type ?? AlertType.info,
    },
  });

  window.dispatchEvent(event);
}
