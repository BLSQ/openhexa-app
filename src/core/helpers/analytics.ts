import { NextRequest } from "next/server";
import getConfig from "next/config";

const { publicRuntimeConfig } = getConfig();

interface TrackEventProperties {
  [key: string]: any;
}

async function sendEvent(
  event: string,
  properties: TrackEventProperties,
  headers?: Headers,
): Promise<void> {
  if (publicRuntimeConfig.DISABLE_ANALYTICS === "true") {
    return;
  }
  const res = await fetch(
    `${publicRuntimeConfig.OPENHEXA_BACKEND_URL ?? ""}/analytics/track/`,
    {
      method: "POST",
      priority: "low",
      headers: headers,
      credentials: "include",
      body: JSON.stringify({ event, properties }),
    },
  );

  if (!res.ok) {
    throw new Error(`Failed to send event: ${res.statusText}`);
  }
}

export async function trackEvent(
  event: string,
  properties: TrackEventProperties,
): Promise<void> {
  if (typeof window === "undefined") {
    // We do not support tracking events on the server
    return;
  }
  try {
    await sendEvent(event, properties);
  } catch (error) {
    console.error("Error tracking event:", error);
  }
}

export async function pageView(request: NextRequest) {
  // This is to be called on the server side only.
  if (typeof window !== "undefined") {
    return;
  }
  try {
    const headers = request.headers;
    headers.set("content-type", "application/json");

    await sendEvent(
      "pageview",
      {
        pathname: request.nextUrl.pathname,
        searchParams: request.nextUrl.searchParams.toString(),
      },
      headers,
    );
  } catch (error) {
    console.error("Error tracking event:", error);
  }
}
