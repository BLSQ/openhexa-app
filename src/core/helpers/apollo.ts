import {
  ApolloClient,
  ApolloLink,
  InMemoryCache,
  NormalizedCacheObject,
  createHttpLink,
  InMemoryCacheConfig,
} from "@apollo/client";
import { onError } from "@apollo/link-error";
import merge from "deepmerge";
import { IncomingHttpHeaders } from "http";
import fetch from "isomorphic-unfetch";
import isEqual from "lodash/isEqual";
import type { AppProps } from "next/app";
import { useMemo } from "react";
import getConfig from "next/config";

const APOLLO_STATE_PROP_NAME = "__APOLLO_STATE__";

export type CustomApolloClient = ApolloClient<NormalizedCacheObject>;

let apolloClient: CustomApolloClient | undefined;

const { publicRuntimeConfig } = getConfig();

const CACHE_CONFIG: InMemoryCacheConfig = {
  // possibleTypes must be provided to cache correctly unions and interfaces
  // https://www.apollographql.com/docs/react/data/fragments/#using-fragments-with-unions-and-interfaces
  possibleTypes: {
    CollectionElement: [
      "DHIS2DataElementCollectionElement",
      "S3ObjectCollectionElement",
    ],
  },
  typePolicies: {
    Team: {
      merge: true,
      fields: {
        permissions: {
          merge: true,
        },
      },
    },
    DAGRun: {
      merge: true,
    },
    User: {
      merge: true,
    },
    Country: {
      // Country code are unique (at least it should). Let's use that for the cache key
      keyFields: ["code"],
    },
    CatalogEntryType: {
      merge: true,
    },
  },
};

const createApolloClient = (headers: IncomingHttpHeaders | null = null) => {
  const enhancedFetch = (url: RequestInfo, init: RequestInit) => {
    return fetch(url, {
      ...init,
      headers: {
        ...init.headers,
        cookie: headers?.cookie ?? "",
        accept: "application/json",
      },
    });
  };

  const link = ApolloLink.from([
    onError(({ graphQLErrors, networkError }) => {
      if (graphQLErrors) {
        graphQLErrors.forEach(({ message, locations, path }) =>
          console.error(
            `[GraphQL error]: Message: ${message}, Location: ${JSON.stringify(
              locations
            )}, Path: ${path}`
          )
        );
      }
      if (networkError) {
        console.error(
          `[Network error]: ${networkError}. Backend is unreachable. Is it running?`
        );
      }
    }),

    createHttpLink({
      uri: (operation) =>
        operation.operationName
          ? `${publicRuntimeConfig.GRAPHQL_ENDPOINT}${operation.operationName}/`
          : publicRuntimeConfig.GRAPHQL_ENDPOINT,
      fetch: enhancedFetch,
      credentials: "include",
      fetchOptions: {
        mode: "cors",
      },
    }),
  ]);

  const cache = new InMemoryCache(CACHE_CONFIG);

  return new ApolloClient({
    ssrMode: typeof window === "undefined",
    ssrForceFetchDelay: 100, // in milliseconds
    link,
    cache,
  });
};

type InitialState = NormalizedCacheObject | undefined;
export interface GetApolloClient {
  headers?: IncomingHttpHeaders | null;
  initialState?: InitialState | null;
}

export const getApolloClient = (
  { headers, initialState }: GetApolloClient = {
    headers: null,
    initialState: null,
  }
) => {
  const client = apolloClient ?? createApolloClient(headers);

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // get hydrated here
  if (initialState) {
    // Get existing cache, loaded during client side data fetching
    const existingCache = client.extract();

    // Merge the existing cache into data passed from getStaticProps/getServerSideProps
    const data = merge(initialState, existingCache, {
      // combine arrays using object equality (like in sets)
      arrayMerge: (destinationArray, sourceArray) => [
        ...sourceArray,
        ...destinationArray.filter((d) =>
          sourceArray.every((s) => !isEqual(d, s))
        ),
      ],
    });
    // Restore the cache with the merged data
    client.cache.restore(data);
  }

  // For SSG and SSR always create a new Apollo Client
  if (typeof window === "undefined") {
    return client;
  } else if (!apolloClient) {
    // Create the Apollo Client once in the client
    apolloClient = client;
  }

  return client;
};

export const addApolloState = (client: ApolloClient<NormalizedCacheObject>) => {
  return { props: { [APOLLO_STATE_PROP_NAME]: client.cache.extract() } };
};

export function useApollo(
  pageProps: AppProps<{ [key: string]: any }>["pageProps"]
) {
  const state = pageProps[APOLLO_STATE_PROP_NAME];
  const client = useMemo(
    () => getApolloClient({ initialState: state }),
    [state]
  );
  return client;
}
