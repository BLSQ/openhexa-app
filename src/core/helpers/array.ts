export function ensureArray<T>(maybeArray: T[] | T): T[] {
  if (typeof maybeArray === "undefined") {
    return [];
  } else if (maybeArray === null) {
    return [];
  } else if (Array.isArray(maybeArray)) {
    return maybeArray;
  } else {
    return new Array(maybeArray);
  }
}
