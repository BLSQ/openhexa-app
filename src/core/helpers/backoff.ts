export type BackoffOptions = {
  min: number;
  max: number;
  factor: number;
  jitter: number;
};

class Backoff {
  private _attempts: number;
  private _base: number;
  private _max: number;
  private _factor: number;
  private _jitter: number;

  constructor(options: BackoffOptions) {
    this._base = options.min || 100;
    this._max = options.max || 10000;
    this._factor = options.factor || 2;
    this._jitter =
      options.jitter > 0 && options.jitter < 1 ? options.jitter : 0;
    this._attempts = 0;
  }

  get attempts(): number {
    return this._attempts;
  }

  duration() {
    let duration = this._base * Math.pow(this._factor, this._attempts++);
    if (this._jitter) {
      var rand = Math.random();
      var deviation = Math.floor(rand * this._jitter * duration);
      duration =
        (Math.floor(rand * 10) & 1) == 0
          ? duration - deviation
          : duration + deviation;
    }
    return Math.min(duration, this._max) | 0;
  }

  reset() {
    this._attempts = 0;
  }
}

export default Backoff;
