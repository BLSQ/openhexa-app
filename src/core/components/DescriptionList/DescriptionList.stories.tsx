import { Story, action } from "@ladle/react";
import React from "react";
import Block from "../Block";
import Button from "../Button";
import DescriptionList from "./DescriptionList";
import { DescriptionListDisplayMode } from "./helpers";

export const DescriptionListStory: Story<
  React.ComponentProps<typeof DescriptionList>
> = (props) => (
  <Block className="max-w-4xl p-4">
    <DescriptionList columns={props.columns} displayMode={props.displayMode}>
      <DescriptionList.Item label="First item">
        <span>And the first item value</span>
      </DescriptionList.Item>
      <DescriptionList.Item label="Second item">
        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
          eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad
          minim veniam, quis nostrud exercitation ullamco laboris nisi ut
          aliquip ex ea commodo consequat. Duis aute irure dolor in
          reprehenderit in voluptate velit esse cillum dolore.
        </p>
      </DescriptionList.Item>
      <DescriptionList.Item label="Third item">
        <Button>A button can be displayed</Button>
      </DescriptionList.Item>
    </DescriptionList>
  </Block>
);

DescriptionListStory.storyName = "default";
DescriptionListStory.args = {
  columns: 1,
};
DescriptionListStory.argTypes = {
  displayMode: {
    options: [
      DescriptionListDisplayMode.LABEL_ABOVE,
      DescriptionListDisplayMode.LABEL_LEFT,
    ],
    control: { type: "radio" },
    defaultValue: DescriptionListDisplayMode.LABEL_LEFT,
  },
};

const defaults = { title: "UI / DescriptionList" };

export default defaults;
