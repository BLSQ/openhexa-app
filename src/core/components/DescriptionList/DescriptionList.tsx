import { InformationCircleIcon } from "@heroicons/react/20/solid";
import clsx from "clsx";
import React, { ReactNode } from "react";
import Tooltip from "../Tooltip";
import { ctx, DescriptionListDisplayMode, useDescriptionList } from "./helpers";

const COLUMNS = {
  1: "",
  2: "grid-cols-2",
  3: "grid-cols-3",
  4: "grid-cols-4",
};

export type DescriptionListProps = React.HTMLAttributes<HTMLDataListElement> & {
  displayMode?: DescriptionListDisplayMode;
  columns?: keyof typeof COLUMNS;
};

const DescriptionList = ({
  children,
  className,
  columns = 1,
  displayMode = DescriptionListDisplayMode.LABEL_LEFT,
}: DescriptionListProps) => {
  return (
    <ctx.Provider value={{ displayMode }}>
      <dl className={clsx("grid gap-6", COLUMNS[columns], className)}>
        {children}
      </dl>
    </ctx.Provider>
  );
};

DescriptionList.Item = function Item({
  children,
  label,
  help,
  className,
}: {
  label: ReactNode;
  children: ReactNode;
  help?: ReactNode;
  className?: string;
}) {
  const { displayMode } = useDescriptionList();
  return (
    <div
      className={clsx(
        displayMode === DescriptionListDisplayMode.LABEL_LEFT &&
          "grid grid-cols-5 gap-2 sm:gap-4",
        displayMode === DescriptionListDisplayMode.LABEL_ABOVE && "space-y-1",
        className
      )}
    >
      <dt
        className={clsx(
          "text-sm font-medium text-gray-500",
          displayMode === DescriptionListDisplayMode.LABEL_LEFT && "col-span-1",
          displayMode === DescriptionListDisplayMode.LABEL_ABOVE && "col-span-5"
        )}
      >
        <span>{label}</span>
        {help && (
          <Tooltip
            placement="top"
            renderTrigger={(ref) => (
              <span ref={ref}>
                <InformationCircleIcon className="ml-1 h-3 w-3 cursor-pointer" />
              </span>
            )}
            label={help}
          />
        )}
      </dt>
      <dd className="col-span-4 text-sm text-gray-900">{children}</dd>
    </div>
  );
};

export default DescriptionList;
