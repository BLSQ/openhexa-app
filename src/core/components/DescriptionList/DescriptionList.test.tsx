import DescriptionList from "./DescriptionList";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { DescriptionListDisplayMode } from "./helpers";

describe("DescriptionList", () => {
  it("renders a list with one item", async () => {
    const { container } = render(
      <DescriptionList>
        <DescriptionList.Item label="label">value</DescriptionList.Item>
      </DescriptionList>
    );

    expect(screen.getByText("label")).toBeInTheDocument();
    expect(screen.getByText("value")).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it("renders the list on multiple columns", () => {
    const { container } = render(
      <DescriptionList columns={2}>
        <DescriptionList.Item label="label 1">value 1</DescriptionList.Item>
        <DescriptionList.Item label="label 2">value 2</DescriptionList.Item>
      </DescriptionList>
    );

    const element = container.getElementsByTagName("dl")[0];

    expect(element).toHaveClass("grid-cols-2");
    expect(container).toMatchSnapshot();
  });

  it("renders the list with the label above the value", () => {
    const { container } = render(
      <DescriptionList
        columns={2}
        displayMode={DescriptionListDisplayMode.LABEL_ABOVE}
      >
        <DescriptionList.Item label="label 1">value 1</DescriptionList.Item>
        <DescriptionList.Item label="label 2">value 2</DescriptionList.Item>
      </DescriptionList>
    );

    const element = container.getElementsByTagName("dl")[0];

    expect(element).toHaveClass("grid-cols-2");

    const items = container.getElementsByTagName("dt");
    Array.from(items).forEach((item) => {
      expect(item).toHaveClass("col-span-5");
    });
    expect(container).toMatchSnapshot();
  });
});
