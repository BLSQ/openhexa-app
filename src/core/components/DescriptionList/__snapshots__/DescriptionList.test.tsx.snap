// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DescriptionList renders a list with one item 1`] = `
<div>
  <dl
    class="grid gap-6"
  >
    <div
      class="grid grid-cols-5 gap-2 sm:gap-4"
    >
      <dt
        class="text-sm font-medium text-gray-500 col-span-1"
      >
        <span>
          label
        </span>
      </dt>
      <dd
        class="col-span-4 text-sm text-gray-900"
      >
        value
      </dd>
    </div>
  </dl>
</div>
`;

exports[`DescriptionList renders the list on multiple columns 1`] = `
<div>
  <dl
    class="grid gap-6 grid-cols-2"
  >
    <div
      class="grid grid-cols-5 gap-2 sm:gap-4"
    >
      <dt
        class="text-sm font-medium text-gray-500 col-span-1"
      >
        <span>
          label 1
        </span>
      </dt>
      <dd
        class="col-span-4 text-sm text-gray-900"
      >
        value 1
      </dd>
    </div>
    <div
      class="grid grid-cols-5 gap-2 sm:gap-4"
    >
      <dt
        class="text-sm font-medium text-gray-500 col-span-1"
      >
        <span>
          label 2
        </span>
      </dt>
      <dd
        class="col-span-4 text-sm text-gray-900"
      >
        value 2
      </dd>
    </div>
  </dl>
</div>
`;

exports[`DescriptionList renders the list with the label above the value 1`] = `
<div>
  <dl
    class="grid gap-6 grid-cols-2"
  >
    <div
      class="space-y-1"
    >
      <dt
        class="text-sm font-medium text-gray-500 col-span-5"
      >
        <span>
          label 1
        </span>
      </dt>
      <dd
        class="col-span-4 text-sm text-gray-900"
      >
        value 1
      </dd>
    </div>
    <div
      class="space-y-1"
    >
      <dt
        class="text-sm font-medium text-gray-500 col-span-5"
      >
        <span>
          label 2
        </span>
      </dt>
      <dd
        class="col-span-4 text-sm text-gray-900"
      >
        value 2
      </dd>
    </div>
  </dl>
</div>
`;
